warehouse/query-core/src/main/java/datawave/query/jexl/visitors/PushdownUnindexedFieldsVisitor.java

```diff
@@ -0,0 +1,171 @@
+    }
+
+    /**
+     * helper method that constructs and applies the visitor.
+     *
+     * @param config
+     *            a config
+     * @param unindexedFields
+     *            the fields considered unindexed
+     * @param queryTree
+     *            the query tree
+     * @param helper
+     *            the metadata helper
+     * @param dataTypeFilter
+     *            the data type filter
+     * @param <T>
+     *            type of the query tree
+     * @return a reference to the node
+     */
+    public static <T extends JexlNode> T pushdownPredicates(T queryTree, ShardQueryConfiguration config, Set<String> unindexedFields, MetadataHelper helper,
+                    Set<String> dataTypeFilter) {
+        PushdownUnindexedFieldsVisitor visitor = new PushdownUnindexedFieldsVisitor(config, unindexedFields, helper, dataTypeFilter);
+        return (T) (queryTree.jjtAccept(visitor, null));
+    }
+
+    @Override
+    public Object visit(ASTAndNode node, Object data) {
+        // if not already delayed somehow
+        if (!QueryPropertyMarker.findInstance(node).isAnyTypeExcept(BOUNDED_RANGE)) {
+            LiteralRange range = JexlASTHelper.findRange().getRange(node);
+
+            if (range != null) {
+                return delayBoundedIndexHole(range, node, data);
+            } else {
+                JexlNode andNode = JexlNodes.newInstanceOfType(node);
+                JexlNodes.copyIdentifierOrLiteral(node, andNode);
+                andNode.jjtSetParent(node.jjtGetParent());
+
+                // We have no bounded range to replace, just proceed as normal
+                JexlNodes.ensureCapacity(andNode, node.jjtGetNumChildren());
+                for (int i = 0; i < node.jjtGetNumChildren(); i++) {
+                    JexlNode newChild = (JexlNode) node.jjtGetChild(i).jjtAccept(this, data);
+                    andNode.jjtAddChild(newChild, i);
+                    newChild.jjtSetParent(andNode);
+                }
+                return andNode;
+            }
+        } else {
+            return node;
+        }
+    }
+
+    /**
+     * Delay the ranges that overlap holes. The range map is expected to only be indexed ranges.
+     *
+     * @param data
+     *            the node data
+     * @param range
+     *            the range
+     * @param currentNode
+     *            the current node
+     * @return a jexl node
+     */
+    protected JexlNode delayBoundedIndexHole(LiteralRange range, ASTAndNode currentNode, Object data) {
+
+        if (missingIndexRange(range)) {
+            return QueryPropertyMarker.create(currentNode, INDEX_HOLE);
+        } else {
+            return currentNode;
+        }
+
+    }
+
+    @Override
+    public Object visit(ASTReferenceExpression node, Object data) {
+        // if not already delayed somehow
+        if (!QueryPropertyMarker.findInstance(node).isAnyTypeExcept(BOUNDED_RANGE)) {
+            return super.visit(node, data);
+        }
```

Comment:
Marker nodes are handled at the level of an ASTAndNode, you are not guaranteed to have a reference expression. You already handle this case above, so this method isn't necessary

Reply:
you are probably correct.  I will validate.